cmd/mt-parrot/monitor.go

```diff
@@ -0,0 +1,154 @@
+	nans             int32   // the partition currently being checked - nope?
+	deltaSum         float64 // sum of abs(value - ts) across the time series
+	numNonMatching   int32   // number of timestamps where value != ts
+	lastSeen         uint32  // the last seen non-NaN time stamp (useful for lag)
+	correctNumPoints bool    // whether the expected number of points were received
+	correctAlignment bool    // whether the last ts matches `now`
+	correctSpacing   bool    // whether all points are sorted and 1 period apart
+}
+
+type partitionMetrics struct {
+	nanCount         *stats.Gauge32 // the number of missing values for each series
+	lag              *stats.Gauge32 // time since the last value was recorded
+	deltaSum         *stats.Gauge32 // the total amount of drift between expected value and actual values
+	nonMatching      *stats.Gauge32 // total number of entries where drift occurred
+	correctNumPoints *stats.Bool    // whether the expected number of points were received
+	correctAlignment *stats.Bool    // whether the last ts matches `now`
+	correctSpacing   *stats.Bool    // whether all points are sorted and 1 period apart
+}
+
+func monitor() {
+	initMetricsBySeries()
+	for tick := range clock.AlignedTickLossless(queryInterval) {
+
+		query := graphite.ExecuteRenderQuery(buildRequest(tick))
+		if query.HTTPErr != nil {
+			httpError.Inc()
+			continue
+		}
+		if query.DecodeErr != nil {
+			decodeError.Inc()
+			continue
+		}
+
+		for _, s := range query.Decoded {
+			processPartitionSeries(s, tick)
+		}
+	}
+}
+
+func processPartitionSeries(s graphite.Series, now time.Time) {
+	partition, err := strconv.Atoi(s.Target)
+	if err != nil {
+		log.Debug("unable to parse partition", err)
+		invalidError.Inc()
+		return
+	}
+	if len(s.Datapoints) < 2 {
+		log.Debugf("partition has invalid number of datapoints: %d", len(s.Datapoints))
+		invalidError.Inc()
+		return
+	}
+
+	serStats := seriesStats{}
+	serStats.lastTs = s.Datapoints[len(s.Datapoints)-1].Ts
+	serStats.correctAlignment = int64(serStats.lastTs) == now.Unix()
+	serStats.correctNumPoints = len(s.Datapoints) == int(lookbackPeriod/testMetricsInterval)+1
+	serStats.correctSpacing = checkSpacing(s.Datapoints)
+
+	for _, dp := range s.Datapoints {
+		if math.IsNaN(dp.Val) {
+			serStats.nans += 1
+			continue
+		}
+		serStats.lastSeen = dp.Ts
+		if diff := dp.Val - float64(dp.Ts); diff != 0 {
+			log.Debugf("partition=%d dp.Val=%f dp.Ts=%d diff=%f", partition, dp.Val, dp.Ts, diff)
+			serStats.deltaSum += diff
+			serStats.numNonMatching += 1
+		}
+	}
+
+	metrics := metricsBySeries[partition]
+	metrics.nanCount.Set(int(serStats.nans))
+	lag := lastPublish - int64(serStats.lastSeen)
+	metrics.lag.Set(int(lag))
+	metrics.deltaSum.Set(int(serStats.deltaSum))
+	metrics.nonMatching.Set(int(serStats.numNonMatching))
+	metrics.correctNumPoints.Set(serStats.correctNumPoints)
+	metrics.correctAlignment.Set(serStats.correctAlignment)
+	metrics.correctSpacing.Set(serStats.correctSpacing)
```

Comment:
there's too many race conditions here.
we update the stats for one partition at a time, rather than one "check" (of all partitions) at a time.
This is exacerbated by the other issue, which is that the reporting of all the stats is scheduled separately from us updating the parameters.
It would be better if we would just update all the stats first, then report them to graphite.
Not sure what's the best way to achieve this, perhaps stats.Graphite could be "triggered" to report rather than having its own reporting ticker.

Reply:
Manually reporting by timestamp in 679eabc714af624fa45c5fea35d3a5956d3a2d39 based on changes in https://github.com/grafana/metrictank/pull/1714